### DISCUSSION TITLE
issue with `next-i18next` Setup in Multi-Tenant App

### DISCUSSION THREAD

**Discussion by yaman3bd** [i18next/next-i18next#2245] [Q&A]
hi, I am working on a Multi-Tenant App, and the content of the tenant is served based on the host, so the host is set as a header in each request sent to the API:
```js
const result = await axios({
  url,
  headers: {
    ...headers,
    //passing the tenant host as a header
    "X-Tenant-Domain": appState.tenantHost
  },
  method,
  data,
  params
});
```
and I am having an issue with how should I set `next-i18next`up. 
I have made the general setup, like setting up `next-i18next` config, wrapping the app in `appWithTranslation`, and using `serverSideTranslations`. but the issue is when too many requests are sent at the time to the server, it does not set the headers at all and the requests are sent from the client, not the server.
here is what my setup looks like:
`_app.tsx`:
```ts
import { appWithTranslation } from 'next-i18next'
import i18nextConfig from "@/next-i18next.config.js";

const MyApp = ({ Component, pageProps }) => (
  <Component {...pageProps} />
)

export default appWithTranslation(MyApp, i18nextConfig)
```
in each page's `getServerSideProps`:
```js
import { serverSideTranslations } from 'next-i18next/serverSideTranslations'

import i18nextConfig from "@/next-i18next.config.js";

export const getServerSideProps = async ({ locale, req }) => {
  const host = req.headers.host;

  //HERE set the custom header for the backend, to the i18next instance
  i18nextConfig.backend.backendOption.customHeaders = {
    "X-Tenant-Domain": host
  };

  return {
    props: {
      ...(await serverSideTranslations(locale, ["common", "footer"], i18nextConfig))
    }
  };
};
```
`next-i18next.config.js`:

```js
const HttpBackend = require("i18next-http-backend/cjs");
const MultiLoadBackendAdapter = require("i18next-multiload-backend-adapter/cjs");
const { axios } = require("./lib/axios");

const env = process.env.NODE_ENV;

module.exports = {
  i18n: {
    locales: ["ar", "en"],
    defaultLocale: "ar"
  },
  ns: [...],
  defaultNS: "common",
  maxParallelReads: 30,
  serializeConfig: false,
  reloadOnPrerender: env === "development",
  use: [MultiLoadBackendAdapter],
  backend: {
    backend: HttpBackend,
    backendOption: {
      loadPath: "{{lng}}|{{ns}}",
      customHeaders: {},
      request: async (options, url, payload, callback) => {
        try {
          const [lng, ns] = url.split("|");

          await axios
            .get("/translations", {
              params: { group: ns.split("+") },
              //passing the tenant host as a header, and here sometimes the headers are empty
              headers: options.customHeaders
            })
            .then((response) => {
              callback(null, {
                data: response.data.data,
                status: 200
              });
            })
            .catch((error) => {
              callback(null, {
                status: 500
              });
            });
        } catch (e) {
          callback(null, {
            status: 500
          });
        }
      }
    }
  }
};
```
and IDK why in some cases the translations are fetched from the client, not the server, and I could not set the headers because I do not have the correct `i18nextConfig` instance to attach the headers to it. 
can anyone help me with this issue to find out what is happening here? I really searched all over Google and made lots of logs and debugging and I could not anything to help resolve the issue.

**Comment by adrai** [comment]
**Possible Root Cause: Shared Singleton Config + Race Condition**

The bug comes down to one thing: **you're mutating a shared singleton**.

In Node.js, `require()` caches module exports. This means every call to `require("@/next-i18next.config.js")` returns the **exact same object in memory**. When you do this in `getServerSideProps`:

```js
i18nextConfig.backend.backendOption.customHeaders = {
  "X-Tenant-Domain": host
};
```

...you're not setting headers for *your* request — you're mutating a global object shared across **all concurrent requests**. Under load, Request A sets the headers, then Request B immediately overwrites them before Request A's translation fetch runs. That's why headers are sometimes empty or wrong.

**Possible Fix: Create a fresh config object per request** (not tested myself)

Never mutate the imported config. Instead, deep-clone or reconstruct it inside `getServerSideProps` for each request:

```js
// getServerSideProps in each page
import { serverSideTranslations } from 'next-i18next/serverSideTranslations'
import baseI18nConfig from "@/next-i18next.config.js";

export const getServerSideProps = async ({ locale, req }) => {
  const host = req.headers.host;

  // Create a NEW config object for this request — do NOT mutate the base config
  const i18nConfig = {
    ...baseI18nConfig,
    backend: {
      ...baseI18nConfig.backend,
      backendOption: {
        ...baseI18nConfig.backend.backendOption,
        customHeaders: {
          "X-Tenant-Domain": host
        }
      }
    }
  };

  return {
    props: {
      ...(await serverSideTranslations(locale, ["common", "footer"], i18nConfig))
    }
  };
};
```

This way, each request should get its own isolated config object with the correct tenant header baked in. (But I've not tested this, you need to check if this really works.)

---

**Why translations sometimes load client-side**

`next-i18next` will fall back to client-side fetching when SSR translation loading fails (e.g., the backend throws or returns 500). When your race condition causes the wrong/empty headers to be used, the backend API likely rejects the request, SSR translation loading fails silently, and the client picks up the job — but at that point there's no `X-Tenant-Domain` header to send.

To guard against silent client-side fallback, make sure your `request` callback surfaces errors clearly:

```js
request: async (options, url, payload, callback) => {
  const [lng, ns] = url.split("|");
  
  // Fail loudly if the tenant header is missing
  if (!options.customHeaders?.["X-Tenant-Domain"]) {
    console.error("[i18n] Missing X-Tenant-Domain header — translations will fail");
    return callback(null, { status: 500 });
  }

  try {
    const response = await axios.get("/translations", {
      params: { group: ns.split("+") },
      headers: options.customHeaders
    });
    callback(null, { data: response.data.data, status: 200 });
  } catch (error) {
    console.error("[i18n] Translation fetch failed:", error.message);
    callback(null, { status: 500 });
  }
}
```
